File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.js
const search = (nums, target) => {
  let c, l = 0, r = nums.length - 1;
  if (nums[0] > nums[r]) {
    while (l < r) {
      c = (l + r) / 2 | 0;
      if (nums[c] > nums[l]) {l = c} else {r = c}
    }
    if (target < nums[0]) {
      ++l;
      r = nums.length - 1;
    } else {
      r = l;
      l = 0;
    }
  }
  if (target < nums[l] || target > nums[r]) return -1;
  while (l < r) {
    c = (l + r) / 2 | 0;
    if (nums[c] < target) {l = c + 1} else {r = c}
  }
  return nums[l] === target ? l : -1;
};

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.go
func search(nums []int, target int) int {
    l, r := 0, len(nums) - 1
    // if array is rotated
    if nums[0] > nums[r] {
      // find rotation point
      for l < r {
        c := (l + r) / 2
        if nums[c] > nums[l] {
          l = c
        } else {
          r = c
        }
      }
      // select search side
      if target < nums[0] {
        l++
        r = len(nums) - 1
      } else {
        r = l
        l = 0
      }
    }
    // perform target search
    if target >= nums[l] && target <= nums[r] {
      for l < r {
        c := (l + r) / 2
        if nums[c] < target {
          l = c + 1
        } else {
          r = c
        }
      }
      if nums[l] == target {
        return l
      }
    }
    return -1
}